File: static/python/lock_ordering_works.py
import time
import threading


class Account:

    _global_id = 0

    def __init__(self, balance):
        self._balance = balance
        self._id = Account._global_id
        Account._global_id += 1
        self._lock = threading.RLock()

    def get_balance(self):
        return self._balance

    @staticmethod
    def transfer(acc_from, acc_to, amount):
        th = threading.Thread(target=acc_from._move, args=(acc_to, amount))
        th.start()

    def _move(self, acc_to, amount):
        # @include
        lock1 = self._lock if self._id < acc_to._id else acc_to._lock
        lock2 = acc_to._lock if self._id < acc_to._id else self._lock
        # Does not matter if lock1 equals lock2: since recursive_mutex locks
        # are reentrant, we will re-acquire lock2.
        with lock1, lock2:
            # @exclude
            if amount > self._balance:
                return False
            acc_to._balance += amount
            self._balance -= amount
            print('returning True')
            return True


def main():
    acc_from = Account(200)
    acc_to = Account(100)
    print('initial balances =', acc_from.get_balance(), acc_to.get_balance())
    Account.transfer(acc_from, acc_to, 50)
    assert (acc_from.get_balance(), acc_to.get_balance()) == (150, 150)
    time.sleep(0.1)
    print('new balances =', acc_from.get_balance(), acc_to.get_balance())
    Account.transfer(acc_from, acc_from, 50)
    assert (acc_from.get_balance(), acc_to.get_balance()) == (150, 150)
    time.sleep(0.1)
    print('new balances =', acc_from.get_balance(), acc_to.get_balance())
    assert (acc_from.get_balance(), acc_to.get_balance()) == (150, 150)


if __name__ == '__main__':
    main()

File: static/cpp/java/src/main/java/com/epi/LockOrderingWorks.java
package com.epi;

public class LockOrderingWorks {
  static class Account {
    private int balance;
    private int id;
    private static int globalId;

    Account(int balance) {
      this.balance = balance;
      this.id = ++globalId;
    }

    private boolean move(Account to, int amount) {
      // @include
      Account lock1 = (id < to.id) ? this : to;
      Account lock2 = (id < to.id) ? to : this;
      synchronized (lock1) {
        // Does not matter if lock1 equals lock2: since Java locks are
        // reentrant, we will re-acquire lock2.
        synchronized (lock2) {
          // @exclude
          if (amount > balance) {
            return false;
          }
          to.balance += amount;
          this.balance -= amount;
          System.out.println("returning true");
          return true;
        }
      }
    }

    public static void transfer(final Account from, final Account to,
                                final int amount) {
      Thread transfer = new Thread(new Runnable() {
        public void run() { from.move(to, amount); }
      });
      transfer.start();
    }
  }

  public static void main(String[] args) throws Exception {
    Account from = new Account(200);
    Account to = new Account(100);
    System.out.println("Initial balances = " + from.balance + " " + to.balance);
    Account.transfer(from, to, 50);
    Thread.sleep(100);
    System.out.println("New balances = " + from.balance + " " + to.balance);
    Account.transfer(from, from, 50);
    Thread.sleep(100);
    System.out.println("New balances = " + from.balance + " " + to.balance);
  }
}
